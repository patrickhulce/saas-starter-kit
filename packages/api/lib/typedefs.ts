import 'express' // tslint:disable-line

declare module 'express-serve-static-core' {
  /* tslint:disable */
  export interface Response {
    body?: any
  }
  /* tslint:enable */
}

export {}
