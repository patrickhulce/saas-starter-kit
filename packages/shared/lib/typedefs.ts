import {ISQLOptions} from 'klay'

export enum ModelID {
  Account = 'account',
  User = 'user',
}

export interface IAccount {
  id?: number
  name: string
  plan: AccountPlan
  createdAt?: Date
  updatedAt?: Date
}

export interface IUser {
  id?: number
  accountId: number
  role: AuthRole
  email: string
  password: string
  firstName: string
  lastName: string
  isVerified: boolean
  verificationKey?: string
  createdAt?: Date
  updatedAt?: Date
}

export enum AuthRole {
  Root = 'root',
  Admin = 'admin',
  User = 'user',
  Anonymous = 'anonymous',
}

export enum Permission {
  RootAccess = '_root_',
  AccountManage = 'account:manage',
  AccountView = 'account:view',
  UserPassword = 'user:password',
  UserProfile = 'user:profile',
  UserView = 'user:view',
}

export enum AccountPlan {
  Free = 'free',
  Bronze = 'bronze',
  Silver = 'silver',
  Gold = 'gold',
}

export interface IConf {
  debug: boolean
  production: boolean
  isUnderTest: boolean

  origin: string
  cdnAppURL: string
  termsOfServiceURL: string
  displayName: string

  secret: string
  database: ISQLOptions

  mailslurp: {apiKey: string}
  sparkpost: {fromAddress: string; apiKey: string; sendToSink: boolean}
}
