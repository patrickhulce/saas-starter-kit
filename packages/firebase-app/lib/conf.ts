import * as functions from 'firebase-functions'
import {IConf} from '../../shared/lib/typedefs'
import conf from '../../shared/lib/conf'

conf.production = !!functions.config().production
conf.apiPathPrefix = '/api'
conf.database.connectionURL = functions.config().sqlURL

export default conf
