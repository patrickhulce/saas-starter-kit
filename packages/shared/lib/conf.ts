import {IConf, JsEnvironment} from './typedefs'

const conf: IConf = {
  debug: true,
  production: false,
  isUnderTest: typeof it === 'function', // tslint:disable-line

  domain: 'THE_DOMAIN',
  displayName: 'THE_PRODUCT_DISPLAY_NAME',
  apiPathPrefix: '/api',

  jsEnvironment: JsEnvironment.Server,
  secret: process.env.SECRET || 'super-secret-string',
  database: {
    connectionURL: process.env.APP_MYSQL_URL || 'mysql://root@localhost/the_product',
  },

  mailslurp: {
    apiKey: process.env.MAILSLURP_API_KEY || '',
  },

  sparkpost: {
    sendToSink: true,
    apiKey: process.env.SPARKPOST_API_KEY || '',
    fromAddress: process.env.SPARKPOST_FROM || '',
  },
}

if (process.env.APP_ENV === 'development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-development') {
  conf.sparkpost.sendToSink = false
}

if (process.env.APP_ENV === 'firebase-production') {
  conf.debug = false
  conf.production = true
  conf.sparkpost.sendToSink = false
}

if (!conf.sparkpost.fromAddress || conf.isUnderTest) {
  conf.sparkpost.fromAddress = `noreply@${conf.domain}`
}

// tslint:disable-next-line
export default conf
