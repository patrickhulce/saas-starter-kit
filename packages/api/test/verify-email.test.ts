import {IState} from './typedefs'
import conf from '../../shared/lib/conf'

describe('change password', () => {
  const state: IState = {}

  beforeAll(() => {
    conf.sparkpost.sendToTestMailboxOnly = false
  })

  afterAll(() => {
    conf.sparkpost.sendToTestMailboxOnly = true
  })

  require('./steps/initialize.test')(state)
  require('./steps/account-setup.test')(state)

  describe('verify email', () => {
    it('should have sent a welcome email', () => {
      expect(__sparkpostSend).toHaveBeenCalled()
      const payload = __sparkpostSend.mock.calls[0][0]
      const address = payload.recipients[0].address
      payload.content.text = payload.content.text.replace(/key=[\w-]+/g, 'key=<key>')
      payload.content.html = payload.content.html.replace(/key=[\w-]+/g, 'key=<key>')
      address.email = address.email.replace(/.*@/, '<uuid>@')
      expect(payload).toMatchSnapshot()
    })

    it('should fetch logged in user', async () => {
      const response = await fetch(`${state.apiURL}/v1/users/me`, {
        headers: {cookie: `token=${state.token}`},
      })

      expect(response.status).toBe(200)
      expect(await response.json()).toEqual(state.user)
    })

    it('should verify email address', async () => {
      let dbUser = await state.userExecutor.findById(state.user.id)
      const fetchOpts = {headers: {cookie: `token=${state.token}`}}
      const query = `key=${dbUser.verificationKey}`
      const response = await fetch(`${state.apiURL}/v1/users/verifications?${query}`, fetchOpts)
      expect(response.url).toMatch(/verification=success/)

      dbUser = await state.userExecutor.findById(state.user.id)
      expect(dbUser.isVerified).toBe(true)
    })
  })

  require('./steps/teardown.test')(state)
})
