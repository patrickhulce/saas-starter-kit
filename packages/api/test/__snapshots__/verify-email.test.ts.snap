// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`change password verify email should have sent a welcome email 1`] = `
Object {
  "content": Object {
    "from": "noreply@THE_DOMAIN",
    "html": "Hello Admin User,<br><br>You're receiving this message because you registered an account for THE_PRODUCT_DISPLAY_NAME.<br>Please follow the link below to verify your email address.<br><br><a href=\\"http://localhost/api/v1/users/verifications?key=<key>\\">http://localhost/api/v1/users/verifications?key=<key></a><br><br>Thanks,<br>The THE_PRODUCT_DISPLAY_NAME Team",
    "subject": "Welcome to THE_PRODUCT_DISPLAY_NAME! - Verify your email",
    "text": "Hello Admin User,

You're receiving this message because you registered an account for THE_PRODUCT_DISPLAY_NAME.
Please follow the link below to verify your email address.

http://localhost/api/v1/users/verifications?key=<key>

Thanks,
The THE_PRODUCT_DISPLAY_NAME Team",
  },
  "options": Object {
    "transactional": true,
  },
  "recipients": Array [
    Object {
      "address": Object {
        "email": "<uuid>@example.com",
        "name": "Admin User",
      },
      "tags": Array [
        "welcome",
        "email-verification",
      ],
    },
  ],
}
`;
